notebooks/Analyze_user_behavior_across_different_lending_protocols.ipynb

```diff
@@ -418,56 +450,134 @@
+    "# Extract unique tokens from the 'Collateral' and 'Debt' columns\n",
+    "collateral_tokens = extract_unique_tokens(df_loans['Collateral'])\n",
+    "debt_tokens = extract_unique_tokens(df_loans['Debt'])\n",
+    "all_tokens = collateral_tokens.union(debt_tokens)\n",
+    "\n",
+    "# Extract token amounts for collateral and debt using the dynamically extracted tokens\n",
+    "collateral_amounts = parse_token_amounts(df_loans['Collateral'], df_loans['Protocol'], all_tokens)\n",
+    "debt_amounts = parse_token_amounts(df_loans['Debt'], df_loans['Protocol'], all_tokens)\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "8b5f526b-4ca2-44e2-9a1c-64bc7be5bac6",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Print the collateral and debt token amounts\n",
+    "print(\"Collateral amounts:\", collateral_amounts)\n",
+    "print(\"Debt amounts:\", debt_amounts)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "109ff2e0-20a2-4026-93fb-fcea1d9651eb",
+   "metadata": {},
+   "source": [
+    "### Aggregating the data"
    ]
   },
   {
    "cell_type": "code",
-   "execution_count": 20,
+   "execution_count": 25,
    "id": "7ac76044-ee76-4807-b497-ad1541ec45a2",
    "metadata": {},
    "outputs": [],
    "source": [
-    "# agregating the data\n",
     "# Convert the aggregated data to DataFrame for better readability\n",
     "collateral_list = [(protocol, token, amount) for protocol, tokens in collateral_amounts.items() for token, amount in tokens.items()]\n",
-    "collateral_df = pd.DataFrame(collateral_list, columns=['Protocol', 'Token', 'Total Collateral (USD)'])\n",
+    "collateral_df = pandas.DataFrame(collateral_list, columns=['Protocol', 'Token', 'Total Collateral (USD)'])\n",
     "\n",
     "debt_list = [(protocol, token, amount) for protocol, tokens in debt_amounts.items() for token, amount in tokens.items()]\n",
-    "debt_df = pd.DataFrame(debt_list, columns=['Protocol', 'Token', 'Total Debt (USD)'])"
+    "debt_df = pandas.DataFrame(debt_list, columns=['Protocol', 'Token', 'Total Debt (USD)'])"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "id": "521fbd91-b7d6-4cfd-99aa-a203e63f3daa",
+   "id": "7beaf9fc-a1e7-4ffc-8e04-83f8aea1e663",
    "metadata": {},
    "outputs": [],
+   "source": []
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "id": "521fbd91-b7d6-4cfd-99aa-a203e63f3daa",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "Protocol        Token \n",
+       "hashstack_v0    DAI       9.689930e+01\n",
+       "                ETH       6.064000e-01\n",
+       "                USDC      1.130584e+03\n",
+       "                USDT      2.268058e+02\n",
+       "                wBTC      2.000000e-03\n",
+       "hashstack_v1    DAI       5.764160e+02\n",
+       "                ETH       1.978970e+01\n",
+       "                Pool      1.595892e+03\n",
```

Comment:
Pool is not a full name of a token, pls adjust the function extracting the tokens.